### DISCUSSION TITLE
Securing API keys for providers?

### DISCUSSION THREAD

**Discussion by nilsbunger** [wevm/wagmi#232] [Q&A] [answered]
What are the best practices for protecting your API keys for providers like Infura and Alchemy? We're building on next.js and the Provider gets added from App.tsx. The naive implementation would make the API key exposed in the client bundle. What do you guys do?

**Comment by tmm** [comment, marked as answer]
Services (like Infura and Alchemy) typically give you publishable and secret keys. Publishable keys that get plugged in to ethers.js Providers are okay to be used on the client side.

With Infura and Alchemy, you can also lock down the origin and/or addresses for the publishable keys if spam is a concern.

**Reply by a-znk** [reply]
I'm curious about the "publishable and secret keys" you reference here. Would the "publishable" key be the API key located in the Alchemy dashboard under APPS-> VIEW KEY-> API KEY?

The safety of exposing these keys is a bit confusing for me. I'm using NEXT_PUBLIC on my key, and it's making me nervous about security, but I can't get the app functioning properly any other way.

How concerned should I be?

**Reply by tmm** [reply]
Like most security questions, the answer is: It depends. There's a trade-off between the convenience of making RPC requests from the browser and locking down access to the RPC service's API key. If you check out the network developer tool for many popular sites (Alchemy, Foundation, etc.), you can see they are calling a RPC provider with their API key.

**Reply by a-znk** [reply]
Super interesting. I pulled screenshots from Foundation and then the Invisible Friends site as a different example. To be super clear, in both instances, the circled yellow keys are exactly what we are talking about here, correct?

(i.e. they are both totally visible in both instances)
![Invisible Friends](https://user-images.githubusercontent.com/105122851/203549886-90b7edb7-36d7-4e90-9c72-ec2998a1f1a6.png)
![Foundation](https://user-images.githubusercontent.com/105122851/203549888-54f1af55-83c4-4423-811a-b624ae0c8be3.png)

**Reply by rsodre** [reply]
Yes, those are the keys.
On next.js, I prefer to export explicitly on `next.config.js` instead of adding `NEXT_PUBLIC`.

```
// next.config.js
module.exports = {
	env: {
		// these keys are public, need to secure on the dashboard
		// https://app.infura.io/dashboard/
		// https://dashboard.alchemy.com/
		INFURA_API_KEY: process.env.INFURA_API_KEY,
		ALCHEMY_API_KEY: process.env.ALCHEMY_API_KEY,
	},
...
}
```

**Reply by a-znk** [reply]
Thanks! Any particular reason you prefer the next.config.js option?

Also, I see "these keys are public, need to secure on the dashboard". I am using Alchemy as my provider, which has an allowlist option. They recommended I use my site's IP address (rather than URL) for the allowlist, because I was getting really inconsistent results with my API calls when using the URL. However, my site is deployed using Vercel, and they don't offer static IP addresses. Therefore, the key is just out there without any "protections" from what I can tell.

Any thoughts on this in your own experience? Perhaps a more suitable place to deploy?

**Reply by rsodre** [reply]
> Thanks! Any particular reason you prefer the next.config.js option?

For organization and readability. I know all the envs that are public just by looking the this file, and don't need to change actual envs on the server if I need to change visibility.

I'm on Vercel too, and use domain names on both.

**Reply by a-znk** [reply]
That makes a lot of sense. I may try that out.

When i use the domain names in the allowlist I get "Unexpected end of JSON input" in the console, and a status code 500 when making my API call about 50% of the time. No issues when i remove from the allowlist. Alchemy's solution was IP allowlist, but can't do with the dynamic Vercel IPs. You ever encounter something like this?

**Reply by rsodre** [reply]
seems to be working fine here

**Reply by rsodre** [reply]
> When i use the domain names in the allowlist I get "Unexpected end of JSON input" in the console, and a status code 500 when making my API call about 50% of the time. No issues when i remove from the allowlist. Alchemy's solution was IP allowlist, but can't do with the dynamic Vercel IPs. You ever encounter something like this?

I just launched my dapp on Vercel 2 days ago and found a ton of problems on mainnet, with API routes crashing and randomly disconnecting users. The log errors were kind of obscure, but they all referenced ethersproject and some of them, Alchemy. Spent yesterday crazy looking for the cause, then I remembered this conversation and removed my domain from the Alchemy allowlist and everything is normal! I will not care about this for a while

Even the days-long continuous rain passed and now it is sunny again. ☀️ 😎

**Reply by a-znk** [reply]
"glad" to hear I'm not the only one... but sorry to hear about the problems! Please do let me know if you encounter a solution.

**Reply by rsodre** [reply]
I think it would be a good idea to report to Vercel and/or Alchemy...
Vercel is clearly running our apis on a different domain.
And we can't see which domains are being blocked on Alchemy.

**Reply by a-znk** [reply]
Could this be different than the "deployment" domain, as opposed to the custom domain? Or are you suggesting it's an entirely different domain hidden to us? I have tried allowlisting both, and still ran into issues... so perhaps that answers my own question.

**Reply by rsodre** [reply]
They actually assign more domains on each deployment...
https://vercel.com/docs/concepts/deployments/generated-urls

**Reply by hjoelh** [reply]
Might be nice to add a note in the docs about this (using provider allowlists to secure your keys)? I'd be happy to open a PR

**Comment by a-znk** [comment]
Oh, intersting. Have you tried adding all of them to the allow list at once?

**Reply by WillyEth** [reply]
Anyone figure this out.  Only way I could get this working with Next.JS was to whitelist on Alchemly,  I deployed on Vercel

**Comment by a-znk** [comment]
I also used the allowlist feature on Alchemy, but it creates a lot of inconsistency for my API routes when I do it this way.

Still working on finding a more reliable solution.

**Reply by WillyEth** [reply]
I been told by a few the only solution is allowlist.  Its required on the frontend

**Comment by a-znk** [comment]
That’s unfortunate. I wish it was more consistent for me.
